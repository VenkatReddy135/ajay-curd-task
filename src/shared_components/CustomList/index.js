/* eslint-disable react/destructuring-assignment */
import React, { memo, useState, useEffect, useCallback } from "react";
import PropTypes from "prop-types";
import CustomButton from "../../shared_components/CustomButton";
import SimpleModal from "../../shared_components/CustomModal";
import CustomToaster from "../../shared_components/CustomToaster";
import CustomInput from "../../shared_components/CustomInput";
import { useNavigate } from "react-router-dom";
import { searchData, getAll, deleteOne } from "../../utils/helperFunction";

const Headers = {
  empName: "Employee Name",
  empAge: "Employee Age",
  empSalary: "Employee Salary",
  actions: "Actions",
};

const labelMapping = {
  NO: "No",
  VIEW: "View",
  YES: "Yes",
};

const TableHeader = memo((props) => {
  return (
    <>
      <th>{Headers.empName}</th>
      <th>{Headers.empAge}</th>
      <th>{Headers.empSalary}</th>
      <th>{Headers.actions}</th>
    </>
  );
});

export default function CustomList(props) {
  const navigate = useNavigate();
  const [employeeData, setEmployeeData] = useState([]);
  const [show, setShow] = useState({ show: false, id: null });
  const [showToast, setShowToast] = useState(false);
  const [successMessage, setSuccessMessage] = useState("");
  const [filteredData, setFilteredData] = useState(employeeData);

  const handleShow = (user) => {
    setShow((prevState) => ({ ...prevState, show: true, id: user.id }));
  };
  const handleClose = () =>
    setShow((prevState) => ({ ...prevState, show: false }));

  const viewEmployeeDetails = (empData) => {
    navigate(`/${empData.id}`);
  };

  const handleSuccess = useCallback(() => {
    let response;
    const clickedId = show.id;
    const deleteCall = async () => {
      response = await deleteOne(props.deleteResource, clickedId);
      setSuccessMessage(response?.data.message);
      setShowToast(true);
    };
    deleteCall();
    setShow((prevState) => ({ ...prevState, show: false }));
  }, [show.id]);

  useEffect(() => {
    const getAllData = async () => {
      const res = await getAll(props.getAllResource);
      setEmployeeData(res.data);
      setFilteredData(res.data);
    };
    getAllData();
  }, []);

  const handleOnChange = (e) => {
    setFilteredData(searchData(e, employeeData));
  };

  return (
    <div className={`container border border-info mt-4 mb-4`}>
      <h2>{props.listTitle}</h2>
      <div className="row">
        <div className="col">
          <CustomInput
            inputType="text"
            placeHolderText="Search"
            inputClass={`form-control searchField marginClass`}
            handleOnChange={handleOnChange}
          />
        </div>
      </div>
      <table className={`table marginClass`}>
        <thead>
          <tr>
            <TableHeader />
          </tr>
        </thead>
        <tbody>
          {filteredData?.length > 0 ? (
            filteredData?.map((user) => (
              <tr key={user?.id}>
                <td>{user?.employee_name}</td>
                <td>{user?.employee_age}</td>
                <td>{user?.employee_salary}</td>
                <td>
                  <CustomButton
                    buttonClass="button btn btn-info"
                    handleClick={() => {
                      viewEmployeeDetails(user);
                    }}
                    buttonValue="View"
                  />
                  <CustomButton
                    buttonClass={`button btn btn-danger marginLeftClass`}
                    handleClick={() => {
                      handleShow(user);
                    }}
                    buttonValue="Delete"
                  />
                </td>
              </tr>
            ))
          ) : (
            <tr>
              <td className="border-0" colSpan={3}>
                {props.textOnNoData}
              </td>
            </tr>
          )}
        </tbody>
      </table>
      <SimpleModal
        showModal={show.show}
        modalContent={props.modalData}
        closeText={labelMapping.NO}
        actionText={labelMapping.YES}
        handleClose={handleClose}
        handleAction={handleSuccess}
      />
      <CustomToaster
        showToast={showToast}
        handleClose={() => setShowToast(false)}
        successMessage={successMessage}
      />
    </div>
  );
}

CustomList.propTypes = {
  listTitle: PropTypes.string,
  textOnNoData: PropTypes.string,
  modalData: PropTypes.string.isRequired,
  getAllResource: PropTypes.any,
  deleteResource: PropTypes.any,
};

CustomList.defaultProps = {
  modalData: "",
};
