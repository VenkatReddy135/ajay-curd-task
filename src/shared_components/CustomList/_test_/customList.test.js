import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import CustomList from "../index";
import { cleanup } from "@testing-library/react";
import { shallow } from "enzyme";
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-17-updated";
Enzyme.configure({ adapter: new Adapter() });

jest.mock("react-router-dom", () => ({
  useNavigate: jest.fn(),
}));
let wrapper;

beforeEach(() => {
  wrapper = shallow(<CustomList />);
});
const props = {
  modalData: "",
};
describe("Custom List Component Testing", () => {
  it("renders component", () => {
    const tree = renderer.create(<CustomList {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<CustomList {...props} />, div);
  });
  it("should have button", () => {
    console.log(wrapper);
    expect(wrapper.find("table")).toHaveLength(1);
  });

  afterEach(() => {
    cleanup();
  });
});
