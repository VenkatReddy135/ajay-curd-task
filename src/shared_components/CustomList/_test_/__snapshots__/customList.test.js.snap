// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom List Component Testing renders component 1`] = `
<div
  className="container border border-info mt-4 mb-4"
>
  <h2 />
  <div
    className="row"
  >
    <div
      className="col"
    >
      <input
        className="form-control searchField marginClass"
        onChange={[Function]}
        placeholder="Search"
        type="text"
      />
    </div>
  </div>
  <table
    className="table marginClass"
  >
    <thead>
      <tr>
        <th>
          Employee Name
        </th>
        <th>
          Employee Age
        </th>
        <th>
          Employee Salary
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="border-0"
          colSpan={3}
        />
      </tr>
    </tbody>
  </table>
  <div
    className="toast-container position-absolute top-0 end-0 p-3"
  />
</div>
`;
